Adds support for fulfillments

Mrkbingham opened the pull request:
# Fulfillments
This merge request seeks to add support for fulfillments and creates models that are reflective of the Square data model.  Orders can have Fulfillments, which themselves have Fulfillments Details of varying types (details on the model update below).  This merge request adds the following:
1. Fulfillment and Fulfillment Details models and migrations
2. Fulfillment and Recipient builders
3. SquareRequest builder calls to support updating of fulfillment details related to orders
4. Tests that cover all the individually added models as well as scenarios for creating and adding orders that contain various Fulfillment Detail types.

## Square API support
This adds support via the `SquareRequestBuilder` to update and modify data on square, and ensure it's synced properly with your application.  Some notable assumptions are baked in here:
1. Orders CANNOT have line-item fulfillment data
2. This also means a single order CANNOT have multiple fulfillments.  This is how the Square UI currently works, so this functionality is only available via the API.  As it's currently an edge case - one that could likely be worked around by creating multiple orders, for now, this will be left out with no plans to add it in.

## Model Updates
This creates a polymorphic relationship between orders via the Fulfillment model.  Adding a fulfillment to an order requires both adding a Fulfillment model and a FulfillmentDetails model - which can be one of three types: `DeliveryDetails`, `PickupDetails` or `ShipmentDetails`.  These relationships are handled through the trait `HasFulfillments`.  

## Recipient Addition
The recipients have been added as a distinct model.  This distinction could be useful for adding future support for additional recipient type, for example [Invoice Recipients](https://github.com/square/square-php-sdk/blob/master/doc/models/invoice-recipient.md).

## Adds Order ID to Payment API request (fulfilling the requirements Square has for orders to show up on the Square Dashboard):
This PR contains a minor fix for adding in the order id to the payments API request, so the payment and order are connected within Square's System.  This update exposed a few bugs that have now been addressed:
1. The Tax calculations in `Util::calculateTotalOrderCost/ByModel` were not taking in to account discounts and inclusive taxes properly.  This now makes sure that Additive taxes are calculated based off the Net Price (taking in to account Discounts and Inclusive Taxes).  Added a private `_calculateNetPrice` method to support this.
2. The Order Builder `buildOrderCopyFromModel` method was not properly creating taxes at the order level.  Added `Constants::DEDUCTIBLE_SCOPE_ORDER`.

Mrkbingham commented:
Also when you get a chance @NikolaGavric94, I'm not able to run StyleCI for this repo, so granting access there would be greatly appreciated so I can clean up any syntax issues: 
![image](https://github.com/NikolaGavric94/laravel-square/assets/4382816/f8e4d06f-fc0b-489f-bd34-1962577920ec)

Mrkbingham committed:
adds item catalog request builder

Mrkbingham committed:
adds tax builder

Mrkbingham committed:
adds catalog image request

Mrkbingham committed:
adds batch upsert catalog method

Mrkbingham committed:
adds create catalog test

Mrkbingham committed:
adds to square facade and class

Mrkbingham committed:
removes ide helper

Mrkbingham committed:
removes call to idehelper

Mrkbingham committed:
test

Mrkbingham committed:
removes testing

Mrkbingham committed:
removes global variable

Mrkbingham committed:
removes ref

Mrkbingham committed:
Merge pull request #1 from Mrkbingham/add_fulfillments

Add fulfillments

Mrkbingham committed:
Merge branch 'master' of https://github.com/Mrkbingham/laravel-square into square_api_updates

Mrkbingham committed:
too many arguments refactoring

Mrkbingham committed:
adds deprecation notice

Mrkbingham committed:
fixes pricing type requirement

Mrkbingham committed:
updates badge

Mrkbingham committed:
Merge pull request #2 from Mrkbingham/square_api_updates

Square api updates

Mrkbingham committed:
makes category id optional for catalog objects

Mrkbingham committed:
updates list catalog test

Mrkbingham committed:
adds batch delete method

Mrkbingham committed:
Merge pull request #3 from Mrkbingham/batch_delete

Batch delete

Mrkbingham committed:
updates hint

Mrkbingham committed:
Merge pull request #4 from Mrkbingham/update_list_catalog_type

updates hint

Mrkbingham committed:
tests batch delete request builder

Mrkbingham committed:
updates syntax

Mrkbingham committed:
updates facade

Mrkbingham committed:
updates reference

Mrkbingham committed:
fixes test

Mrkbingham committed:
Merge pull request #5 from Mrkbingham/batch_delete_builder

Batch delete builder

Mrkbingham committed:
adds location_id to migration

Mrkbingham committed:
adds location_id during order save

Mrkbingham committed:
adds location id to model factory

Mrkbingham committed:
Merge pull request #6 from Mrkbingham/add_location_id

Add location

NikolaGavric94 commented:
Hey Mark,

I will take a detailed look at this over the upcoming weekend and will get it sorted by next Tuesday, thank you for your effort and willingness to help :)

Mrkbingham committed:
adds missing fulfillments collection to array builder

Mrkbingham committed:
Merge pull request #7 from Mrkbingham/add_fulfillments

adds missing fulfillments collection to array builder

Mrkbingham committed:
adds exception if recipient is missing

Mrkbingham committed:
adds exception and test to validate exception

Mrkbingham committed:
Merge pull request #8 from Mrkbingham/add_fulfillments

adds exception if recipient is missing

Mrkbingham committed:
adds validation rules to enforce RFC 3339 format

Mrkbingham committed:
adds missing semicolon

Mrkbingham committed:
adds format type to constants, updates tests

Mrkbingham committed:
reverts test and constant change in favor of serialize date function

Mrkbingham committed:
updates to use constants

Mrkbingham committed:
Merge pull request #9 from Mrkbingham/add_fulfillments

Adds validation rules to fulfillment models to enforce RFC 3339 format

Mrkbingham committed:
fixes products copy/paste reference

Mrkbingham committed:
fixes courier pickup naming convention

Mrkbingham committed:
Merge pull request #10 from Mrkbingham/add_fulfillments

Adds date formatting to square builder, fixes errant fulfillments reference

Mrkbingham committed:
updates to the datetime time stamp reference

Mrkbingham committed:
modifies fulfillment square builder to adhere to single fulfillment syntax

Mrkbingham committed:
Merge pull request #11 from Mrkbingham/add_fulfillments

Updates date format to DateTime::RFC3339, fixes Square build fulfillment error

Mrkbingham committed:
allows placed_at to be nullable

Mrkbingham committed:
Merge pull request #12 from Mrkbingham/add_fulfillments

allows placed_at to be nullable - not to be set before creating the order

Mrkbingham committed:
adds set order id to charge request

Mrkbingham committed:
Merge pull request #13 from Mrkbingham/add_fulfillments

Adds order ID to charge request in request builder

Mrkbingham committed:
adds array_key check

Mrkbingham committed:
merges taxes and discounts into single collections

Mrkbingham committed:
adds product tax calculation

Mrkbingham committed:
updates price on test

Mrkbingham committed:
adds comments and more explicit variables

Mrkbingham committed:
adds product discount test

Mrkbingham committed:
adds order discount test

Mrkbingham committed:
renames discount variable for claity

Mrkbingham committed:
adds net price calculation to account for inclusive tax

Mrkbingham committed:
fixes order builder variable

Mrkbingham committed:
updates test to correct value

Mrkbingham committed:
simplifies total cost calculation

Mrkbingham committed:
simplifies function

Mrkbingham committed:
refactors to simplify

Mrkbingham committed:
simplifies return statement

Mrkbingham committed:
Merge pull request #14 from Mrkbingham/add_fulfillments

adds array_key check

Mrkbingham committed:
adds fulfillment builder to construct

Mrkbingham committed:
Merge pull request #15 from Mrkbingham/add_fulfillments

adds fulfillment builder to construct

Mrkbingham committed:
adds more context to save failures

Mrkbingham committed:
Merge pull request #16 from Mrkbingham/add_fulfillments

Adds more context to save failures

Mrkbingham committed:
adds support for adding recipient data via an array

Mrkbingham committed:
array or null

Mrkbingham committed:
removes extra if check

Mrkbingham committed:
removes unnecessary extra block

Mrkbingham committed:
refactors to simplify

Mrkbingham committed:
refactors detail check into reusable function

Mrkbingham committed:
Merge pull request #17 from Mrkbingham/add_fulfillments

adds support for adding recipient data via an array

NikolaGavric94 commented:
Would you be able to add the wiki page changes with the examples, notes and limitations of using fulfillments with this package in this PR?

NikolaGavric94 commented on tests/unit/DeliveryDetailsTest.php:
```diff
@@ -0,0 +1,46 @@
+        $fulfillment->order()->associate($this->data->order);
+        $fulfillment->save();
+
+        $this->assertInstanceOf(DeliveryDetails::class, $fulfillment->fresh()->fulfillmentDetails);
```
Can we do some negative testing around these models too? This one in particular needs more overall tests

NikolaGavric94 commented on tests/unit/FulfillmentTest.php:
```diff
@@ -0,0 +1,133 @@
+        $order->save();
+
+        $this->assertInstanceOf(Order::class, $fulfillment->order);
+    }
```
Needs more negative tests

Mrkbingham replied:
Added more tests, including some square order tests.

NikolaGavric94 commented on tests/unit/PickupDetailsTest.php:
```diff
@@ -0,0 +1,46 @@
+        $fulfillment->save();
+
+        $this->assertInstanceOf(PickupDetails::class, $fulfillment->fresh()->fulfillmentDetails);
+    }
```
Needs more tests in general with additions of negative tests please

Mrkbingham replied:
Added more tests, along with negative tests.

NikolaGavric94 commented on tests/unit/ShipmentDetailsTest.php:
```diff
@@ -0,0 +1,46 @@
+        $fulfillment->save();
+
+        $this->assertInstanceOf(ShipmentDetails::class, $fulfillment->fresh()->fulfillmentDetails);
+    }
```
Same as for pickupdetailstest class

Mrkbingham replied:
Added more tests here as well!

Mrkbingham committed:
adds back in ref version

Mrkbingham commented on src/builders/FulfillmentBuilder.php:
```diff
@@ -0,0 +1,263 @@
+            throw new InvalidSquareOrderException('Fulfillment type does not match details', 500);
+        }
+
+        $fulfillmentObj = $tempFulfillment;
```
Yeah, that's a good call.  I was copying the format here from the product builder. Not something I think needs changing in this PR, but worth noting that there may be an unnecessary reassignment over here:
https://github.com/NikolaGavric94/laravel-square/blob/master/src/builders/ProductBuilder.php#L167

Mrkbingham replied:
Updated this to use the tempFulfillment throughout the function instead of reassigning here:
https://github.com/Mrkbingham/laravel-square/commit/c295919974780205a92f7116dcf706b628ee707d

Mrkbingham replied:
Minor tweak on this one as the original refactor here wasn't correct: https://github.com/Mrkbingham/laravel-square/commit/ddb8dc6b4e5c172dc6165a00efb169cd6806ed2f

Mrkbingham committed:
removes unneeded reassignment

Mrkbingham commented on src/builders/FulfillmentBuilder.php:
```diff
@@ -0,0 +1,263 @@
+        // Check for recipient
+        $fulfillmentDetailsCopy->recipient = $this->getRecipientFromFulfillmentArray($fulfillment, $type);
+
+        $fulfillmentObj = $tempFulfillment;
```
Updated!  No longer reassigning this - commit for reference here:
https://github.com/Mrkbingham/laravel-square/commit/c295919974780205a92f7116dcf706b628ee707d

Mrkbingham committed:
fixes reassignment issues

Mrkbingham commented on src/builders/OrderBuilder.php:
```diff
@@ -147,6 +152,41 @@ public function buildOrderFromOrderCopy(Model $order, stdClass $orderCopy): Mode
+                    // Create the recipient
+                    $recipient = $fulfillment->fulfillmentDetails->recipient;
+                    $recipient->save();
+                    unset($fulfillment->fulfillmentDetails->recipient);
```
The recipient needs to be unset as the `FulfillmentDetails` model has a many-to-one relationship with the `Recipient` model.  Without unsetting, saving the `FulfillmentDetails` models throws this error:
```
Caused by
PDOException: SQLSTATE[HY000]: General error: 1 table nikolag_shipment_details has no column named recipient
```

I've added a comment to clarify the goal of the `unset`

Mrkbingham replied:
Resolving this one as it's been open for a bit with what I think is a pretty good resolution.

Mrkbingham committed:
adds unset comment

Mrkbingham committed:
removes unnecessary unset

Mrkbingham commented on src/builders/OrderBuilder.php:
```diff
@@ -147,6 +152,41 @@ public function buildOrderFromOrderCopy(Model $order, stdClass $orderCopy): Mode
+                    // Create the fulfillment details
+                    $fulfillment->fulfillmentDetails->save();
+                    $fulfillment->fulfillmentDetails()->associate($fulfillment->fulfillmentDetails);
+                    unset($fulfillment->fulfillmentDetails);
```
This does not need to be unset.  I've removed this:
https://github.com/Mrkbingham/laravel-square/commit/977bbb09fd97521667fe12c8df7a461a4bbac374

Mrkbingham committed:
adds customer model requirements note

Mrkbingham commented on src/builders/RecipientBuilder.php:
```diff
@@ -0,0 +1,82 @@
+        if (!$hasCustomerID) {
+            Validator::make($recipient->toArray(), $individualFieldsRules)->validate();
+        } else {
+            // As long as the customer_id is present, we're good
```
I was basing the required data off of Square's model convention:
https://github.com/square/square-php-sdk/blob/master/doc/models/fulfillment-recipient.md

Their notes say:
The ID of the customer associated with the fulfillment.

> If customer_id is provided, the fulfillment recipient's display_name, email_address, and phone_number are automatically populated from the targeted customer profile.

I've updated the comment here with a reference to that link to add more clarity:
https://github.com/Mrkbingham/laravel-square/commit/3fa664ed21858f4430f1a2aeabb15d13229e0da5

Mrkbingham committed:
removes action again

Mrkbingham commented on .github/workflows/main.yml:
```diff
@@ -38,8 +38,6 @@ jobs:
     steps:
       - uses: actions/checkout@v3
-        with:
-          ref: '3.3.x'
```
I removed this in an effort to run actions on my forked repo.  It looks like there's a branching naming convention I might not have adhered to here, so this does bring up a larger question related to the merge: should I rename my branch to 3.3.1?  I may be misunderstanding how actions work, so if you're aware of any other alternative setups that would help me achieve running tests in my fork I'd be happy to hear of any!  

I'll leave this one open for now in case you've got any suggestions - my intention will be to resolve this issue once all the items below have been taken care of (and I'm able to validate my branch is passing PHPUnit tests successfully) - so long-term I don't intend to keep this change.

Mrkbingham commented on src/builders/SquareRequestBuilder.php:
```diff
@@ -269,6 +366,117 @@ public function buildAppliedTaxes(Collection $taxes): array
+        $fulfillmentDeliveryDetails->setCancelReason($deliveryDetails->cancel_reason);
+        $fulfillmentDeliveryDetails->setRejectedAt($deliveryDetails->rejected_at?->format(Constants::DATE_FORMAT));
+
+        return $fulfillmentDeliveryDetails;
```
Are you referring to accessors on the individual properties?
https://laravel.com/docs/11.x/eloquent-mutators#defining-an-accessor

I definitely really like that approach, just want to confirm.

Mrkbingham committed:
fixes typo

Mrkbingham committed:
cleans up fillable delivery details

Mrkbingham commented on src/models/DeliveryDetails.php:
```diff
@@ -0,0 +1,113 @@
+        'courier_support_phone_number',
+        'square_delivery_id',
+        'external_delivery_id',
+        'managed_deliver',
```
Good call - here's the fields I've removed, lmk what you think:
https://github.com/Mrkbingham/laravel-square/commit/6e573b5846119784cf1ef776558a257a1de42500

Mrkbingham committed:
cleans up pickup details

Mrkbingham commented on src/models/PickupDetails.php:
```diff
@@ -0,0 +1,108 @@
+        'canceled_at',
+        'cancel_reason',
+        'is_curbside_pickup',
+        'curbside_pickup_details',
```
Cleaned up the PickupDetails fillable array here - lmk what you think:
https://github.com/Mrkbingham/laravel-square/commit/797811fc81edc7185cf2c680bb86f68dc3715e64

Mrkbingham committed:
removes customer id from fillable

Mrkbingham commented on src/models/Recipient.php:
```diff
@@ -0,0 +1,55 @@
+     * @var array
+     */
+    protected $fillable = [
+        'customer_id',
```
Removed! https://github.com/Mrkbingham/laravel-square/commit/08276661fac9227baf6474f48763c2dd14abf252

Mrkbingham committed:
cleans up shipment details fillable array

Mrkbingham commented on src/models/ShipmentDetails.php:
```diff
@@ -0,0 +1,103 @@
+     * @var array
+     */
+    protected $fillable = [
+        'recipient_id',
```
Cleaned up here: https://github.com/Mrkbingham/laravel-square/commit/f2901b5907be120167b546565bfe018a35058e14

Lmk what you think

Mrkbingham commented:
> Would you be able to add the wiki page changes with the examples, notes and limitations of using fulfillments with this package in this PR?

Definitely!  I'm heading out for vacation this week, so will get to this when I return, but will be sure to add a wiki page with examples using the existing ones as a reference point.

Mrkbingham committed:
fixes product reference in hasfulfillment

Mrkbingham commented on src/traits/HasFulfillments.php:
```diff
@@ -0,0 +1,38 @@
+    public function hasFulfillment(mixed $fulfillment): bool
+    {
+        $val = is_array($fulfillment)
+            ? (array_key_exists('id', $fulfillment) ? Product::find($fulfillment['id']) : $fulfillment )
```
:facepalm: There was definitely a copy/paste error here.  Fixed here!
https://github.com/Mrkbingham/laravel-square/commit/cfe65f19839203950a38e530e91da8a2d098bd45

Mrkbingham committed:
updates schedule type to use constant reference

Mrkbingham committed:
adds make and create tests for fulfillment models

Mrkbingham committed:
adds no order fulfillment negative test

Mrkbingham committed:
adds thrown exception

Mrkbingham committed:
adds negative fulfillment test

Mrkbingham committed:
Adds additional negative fulfillment tests

Mrkbingham committed:
adds negative association tests

Mrkbingham committed:
fixes test name

Mrkbingham committed:
simplifies make test

Mrkbingham commented:
@NikolaGavric94 After a bit of a long hiatus from looking at this, I've resumed work again!  I've added quite a few more tests along with negative testing coverage.  Happy to add anything more that you think is missing.  I still need to add the Wiki pages, but outside of that I think everything should be ready here (other than the one open thread left above re: `mail.yml`)

Mrkbingham committed:
removes period

Mrkbingham committed:
use statement fixes

Mrkbingham committed:
syntax

Mrkbingham commented:
Okay, I've added the wiki page here:
https://github.com/NikolaGavric94/laravel-square/wiki/Examples-with-Fulfillments-(coming-soon)

Mrkbingham committed:
Merge branch 'add_fulfillments' of https://github.com/Mrkbingham/laravel-square into HEAD

Mrkbingham committed:
adds product syncing

Mrkbingham committed:
Merge pull request #18 from Mrkbingham/add_fulfillments

adds back in ref version

Mrkbingham committed:
simplifies return statement

Mrkbingham committed:
adds required fields validation

Mrkbingham committed:
Merge pull request #19 from Mrkbingham/sync_products

Sync products

Mrkbingham committed:
adds description field to products table

Mrkbingham committed:
Merge branch 'master' of https://github.com/Mrkbingham/laravel-square into add_description_field_to_products

Mrkbingham committed:
syncs description to new desc field

Mrkbingham committed:
adds missing semicolon

Mrkbingham committed:
fixes table name

Mrkbingham committed:
Merge pull request #20 from Mrkbingham/add_description_field_to_products

Adds description field to products table

NikolaGavric94 commented:
I will check the changes this upcoming weekend and we will see where we are at, thanks for all your hard effort @Mrkbingham

Mrkbingham committed:
Merge branch 'add_location_id_to_order_table' of https://github.com/Mrkbingham/laravel-square

Mrkbingham committed:
Merge branch 'master' of https://github.com/Mrkbingham/laravel-square

NikolaGavric94 commented:
Would you be able to just attach the picture of passing tests here and I'll merge it. Need to get on fixing actions when PRs are made from forks

Mrkbingham commented:
FYI @NikolaGavric94 I did make this new MR that contains part of the changes included in here.  I really should have done this sooner, but in order to more easily digest some of the changes, I figured this might be helpful if we ever need to go back and do any debugging:

https://github.com/NikolaGavric94/laravel-square/pull/95

Mrkbingham committed:
Merge branch 'tax_calculation_fixes' of https://github.com/Mrkbingham/laravel-square into add_fulfillments

Mrkbingham committed:
syntax

Mrkbingham committed:
more syntax

Mrkbingham committed:
syntax

Mrkbingham committed:
triggering pipeline

Mrkbingham commented:
Okay, in addition to branching out a subset of this MR - I've also created a "master" branch in my repo called "nikolag_master" that was just synced from the current state of master of this main repo.  I've created MRs in my own repo against that branch, and have linked here to this successful pipeline run:
https://github.com/Mrkbingham/laravel-square/pull/25/checks

There were a few syntax issues, but those are now resolved too!

Mrkbingham committed:
fixes schedule type typo

Mrkbingham committed:
fixes schedule type table reference

Mrkbingham committed:
fixes factory references

Mrkbingham committed:
removes carrier from est

Mrkbingham committed:
fixes fillable

Mrkbingham committed:
refactors to move fulfillment requests builders into separate class

Mrkbingham committed:
fixes class reference

Mrkbingham committed:
adds missing use statements

Mrkbingham committed:
adds a space

Mrkbingham committed:
removes a space

Mrkbingham committed:
fixes php tag

Mrkbingham committed:
Merge branch 'add_fulfillments' of https://github.com/Mrkbingham/laravel-square

Mrkbingham committed:
removes unnecessary columns

Mrkbingham committed:
fixes the fulfillment relationship on save during order create

Mrkbingham committed:
Merge branch 'add_fulfillments' of https://github.com/Mrkbingham/laravel-square

Mrkbingham committed:
Merge branch 'tax_calculation_fixes' of https://github.com/Mrkbingham/laravel-square

Mrkbingham committed:
adds address model to recipient, adds test and adds to square request

Mrkbingham committed:
Merge branch 'add_fulfillments' of https://github.com/Mrkbingham/laravel-square

Mrkbingham committed:
Merge branch 'add_square_api_details_to_exception' of https://github.com/Mrkbingham/laravel-square

Mrkbingham committed:
syntax

Mrkbingham committed:
table updates

Mrkbingham committed:
Merge branch 'add_fulfillments' of https://github.com/Mrkbingham/laravel-square

Mrkbingham committed:
reverts order id back to string

Mrkbingham committed:
Merge branch 'add_fulfillments' of https://github.com/Mrkbingham/laravel-square

Mrkbingham committed:
fixes errant change to pivot type column

Mrkbingham committed:
Merge branch 'add_fulfillments' of https://github.com/Mrkbingham/laravel-square

Mrkbingham committed:
adds fulfillment uid to pickup details table, aligns indexes

Mrkbingham committed:
Merge branch 'add_fulfillments' of https://github.com/Mrkbingham/laravel-square

Mrkbingham committed:
schedule type, fulfillment and recipient id column length limits

Mrkbingham committed:
Merge branch 'add_fulfillments' of https://github.com/Mrkbingham/laravel-square

Mrkbingham committed:
fixes column timestamp typo

Mrkbingham committed:
Merge branch 'add_fulfillments' of https://github.com/Mrkbingham/laravel-square

Mrkbingham committed:
fixes model casting reference

Mrkbingham committed:
performance tuning of recipients table and adds square customer id

Mrkbingham committed:
Merge branch 'add_fulfillments' of https://github.com/Mrkbingham/laravel-square

Mrkbingham committed:
adds catalog object id to table

Mrkbingham committed:
adds to fillable

Mrkbingham committed:
fixes column drop

Mrkbingham committed:
fixes service and test

Mrkbingham committed:
adds catalog object id to order build

Mrkbingham committed:
Merge branch 'sync_products' of https://github.com/Mrkbingham/laravel-square

Mrkbingham committed:
adds square id to taxes

Mrkbingham committed:
Merge branch 'add_taxes_square_cataolog_object_id' of https://github.com/Mrkbingham/laravel-square

Mrkbingham committed:
updates to use variation id

Mrkbingham committed:
Merge branch 'sync_products' of https://github.com/Mrkbingham/laravel-square

Mrkbingham committed:
adds square_catalog_object_id to builders

Mrkbingham committed:
Merge branch 'add_taxes_square_cataolog_object_id' of https://github.com/Mrkbingham/laravel-square

Mrkbingham committed:
Merge branch 'add_taxes_square_cataolog_object_id' of https://github.com/Mrkbingham/laravel-square into sync_products

Mrkbingham committed:
adds square dates to tables

Mrkbingham committed:
copy paste fix

Mrkbingham committed:
updates fillables

Mrkbingham committed:
adds discount syncing test

Mrkbingham committed:
adds square tax sync

Mrkbingham committed:
Merge branch 'sync_products' of https://github.com/Mrkbingham/laravel-square

Mrkbingham committed:
fixes errant merge reference

NikolaGavric94 commented:
@Mrkbingham Would you be able to resolve these conflicts?

Mrkbingham committed:
Merge branch 'master' of https://github.com/NikolaGavric94/laravel-square into add_fulfillments

Mrkbingham commented:
@NikolaGavric94 resolved!

Mrkbingham committed:
removes tracking of vscode files

Mrkbingham committed:
adds dump to debug test failure

Mrkbingham committed:
reverts debugging

Mrkbingham commented:
@NikolaGavric94 I'm also noting here that in my repo, I've got the env variables defined in order to get the PHPUnit tests passing, and it looks like those are not defined here:
https://github.com/NikolaGavric94/laravel-square/wiki/Installation-guide

I did a little bit of digging before realizing that the tests here are not likely to pass, however, they're passing over here on the replica PR in my fork:
https://github.com/Mrkbingham/laravel-square/pull/25/checks

Mrkbingham committed:
adds morphs to include index

Mrkbingham committed:
uses model constants

Mrkbingham committed:
fixes foreign id column type

Mrkbingham committed:
adds model constants for delivery details table

Mrkbingham committed:
adds constants for the pickup details table

Mrkbingham committed:
adds foreign ids

Mrkbingham committed:
comments

Mrkbingham committed:
fixes customer foreign id

Mrkbingham committed:
pickup details review

Mrkbingham committed:
delivery table review

Mrkbingham committed:
use square constants for fulfillment type

Mrkbingham committed:
uses square type constants

Mrkbingham committed:
makes foreign id nullable again

Mrkbingham committed:
removes unnecessary index

Mrkbingham committed:
modifies indices

Mrkbingham committed:
integrates index additions with original table create methods

Mrkbingham committed:
Merge branch 'master' into add_fulfillments

Mrkbingham committed:
removes orders foreign id, adds comment in case

Mrkbingham committed:
reverts foreign id change

Mrkbingham committed:
removes out-of-order table references

Mrkbingham committed:
makes customers nullable

Mrkbingham committed:
adds index name to prevent long index name error

Mrkbingham committed:
unsigned

Mrkbingham committed:
changes column type to string 375

Mrkbingham committed:
reverts pickup details changes

NikolaGavric94 reviewed (commented):
@Mrkbingham Can we revisit `fulfillments` with the new version `40` square once that PR goes in?

Mrkbingham commented:
@NikolaGavric94 Yeah, since the last round of reviews, I've added a handful of changes - primarily to improve DB performance with tables that have a lot of fulfillments in them, along with a few minor bug-fixes.  I definitely think the Laravel 11 and Square library v 40 make the most sense to move in first before spending too much time on any of the other open PRs.

Mrkbingham committed:
Merge branch 'master' of https://github.com/NikolaGavric94/laravel-square into add_fulfillments

Mrkbingham committed:
Merge branch 'master' of https://github.com/NikolaGavric94/laravel-square into add_fulfillments

Mrkbingham committed:
adds fulfillment_date attribute accessor

Mrkbingham committed:
removes date attribute getter

Mrkbingham committed:
adds missing options to fillable arrays

Mrkbingham committed:
adds digital and other types

Mrkbingham committed:
reverts fulfillment change

Mrkbingham committed:
reverts change to readme, adds me as contributor

Mrkbingham committed:
reverts composer.json changes

Mrkbingham committed:
Merge branch 'master' of https://github.com/NikolaGavric94/laravel-square into add_fulfillments

Mrkbingham committed:
fixes merge issue

Mrkbingham committed:
syntax

Mrkbingham committed:
syntax

Mrkbingham committed:
branch cleanup

Mrkbingham committed:
removes unused use

Mrkbingham committed:
removes unused constant

Mrkbingham committed:
type hints

Mrkbingham committed:
reducing git diffs

Mrkbingham committed:
syntax

Mrkbingham committed:
phpdoc block fixes

Mrkbingham committed:
adds use statement

Mrkbingham committed:
doc block fixes

Mrkbingham committed:
updates comments

Mrkbingham committed:
spacing

Mrkbingham committed:
removes comma from test

Mrkbingham committed:
adds fulfillment test

Mrkbingham committed:
updates square facade

Mrkbingham committed:
doc block fixes

Mrkbingham commented:
@NikolaGavric94 oh great call - yeah I went ahead and updated the docs over here.  They did have an errant reference to a constant property that was removed.  I've updated the [Order Fulfillment doc](https://github.com/NikolaGavric94/laravel-square/wiki/Order-Fulfillments), the [Webhooks doc](https://github.com/NikolaGavric94/laravel-square/wiki/Webhooks) and the [Square Facade](https://github.com/NikolaGavric94/laravel-square/wiki/Square-Facade) - as well as making note of all of the new pages on the home page.

There were also some missing methods from the Square facade, so I've added them to this PR.

Mrkbingham committed:
makes uid unique

Mrkbingham committed:
adds fulfillment id to fulfillment during building to enforce cascade on deletion

Mrkbingham committed:
updates recipient to one-to-one relationship

Mrkbingham commented:
@NikolaGavric94 While implementing this I ran into a data integrity issue that I've resolved by modifying the relationship between fulfillments and recipients.  In short, recipients have a one-to-one relationship with fulfillments, and not a one-to-many as it had previously been written.  All the test are passing, but in case you're curious to review this change I've limited it to a single commit here:
https://github.com/NikolaGavric94/laravel-square/pull/88/commits/347b742a91904681569c2b5f414bc6f2d9ce5491

Mrkbingham committed:
fixed an issue with recipient creation from array

Mrkbingham committed:
adds missing recipient exception

Mrkbingham committed:
reverts recipient requirement changes

Mrkbingham committed:
minor cleanup

Mrkbingham committed:
adds deliver at to fillable

Mrkbingham committed:
Merge branch 'master' of https://github.com/NikolaGavric94/laravel-square into add_fulfillments

Mrkbingham commented:
@NikolaGavric94 didn't realize I left this one as draft - marking it ready now!

Mrkbingham committed:
Syntax

Add a comment about fulfillment limitations in the API.

Mrkbingham committed:
syntax
